Add trending links

Gargron opened the pull request:
### Changes to trending hashtags in the admin UI

Rework the admin UI around trending hashtags. The list of all hashtags and the different (but unclear) ways of sorting them were impractical. Since hashtags are only meaningful in so far as they are being actively used, the list of hashtags has been moved to the "Trends" section and is now populated from trending hashtags only, sorted according to use. Unlike previously, the system now keeps track of trending hashtags even for those of them that have not been approved or have been rejected, allowing you to see which hashtags are nevertheless in use at the moment.

![grafik](https://user-images.githubusercontent.com/184731/141406599-1d1b109e-f763-44a7-be93-e3aefb6abb00.png)

The list also gives more information on how exactly the hashtags are trending. So, you can see the rank - which position in the publicly visible trending section the hashtag has - but also the underlying score, and when the score peaked. Individual hashtag pages have received a long overdue facelift, making use of the metrics APIs and components introduced in #16829, allowing you to see the use history over time as well as a list of languages and servers in the hashtag.

![grafik](https://user-images.githubusercontent.com/184731/141406418-e0f61601-9131-4166-ac96-10df733cc085.png)

The e-mail notifications sent to the staff about new hashtags requiring review have also been changed. Instead of sending one e-mail per hashtag, one e-mail is sent for a whole batch. The e-mail explains, for each hashtag, why it is considered as a trending candidate: how many times it was used today, vs yesterday, and the resulting score. It also tells you which is the lowest currently visible trending hashtag and its score - it is that score that each of the candidates has beat.

### Trending links

Trending links are a discovery feature. While trending hashtags shows which hashtags are being used (and shared through reblogs) at the moment, trending links does the same for links, providing a sort of link-aggregator-esque view of the fediverse, allowing you to find interesting stories that people talk about. Unlike trending hashtags which require the admin to approve every hashtag individually before it can appear publicly, in this case only the publisher of a link (i.e. the domain of the link) must be approved (e.g. nytimes.com).

The admin interfaces provides a view of the trending links with a similar level of detail as for hashtags.

![grafik](https://user-images.githubusercontent.com/184731/141407623-700c7a2c-5304-4830-addb-a6c31600f2d7.png)

Through another page, publishers can be managed directly. Publishers affect subdomains as well.

![grafik](https://user-images.githubusercontent.com/184731/141407656-a7204867-55ea-4520-aca3-c6029b877b46.png)

Mock-up of the public-facing feature in the official iOS app:

![iPhone Trends - Links](https://user-images.githubusercontent.com/184731/139561392-e8666124-4977-4121-89fc-7aa7201eccea.png)

### REST API changes

- `GET /api/v1/trends` is deprecated, use `GET /api/v1/trends/tags` instead
- `GET /api/v1/trends/links` added
- `GET /api/v1/admin/trends` is deprecated, use `GET /api/v1/admin/trends/tags`

### Other changes

- Previously, a hashtag review request was sent only when a hashtag was about to enter the top 3 spots on the list. This number has been changed to 10, since 10 is the number of results returned from the public API.

Gargron committed:
Add trending links

Gargron committed:
Add overriding specific links trendability

ClearlyClaire commented on app/models/trends.rb:
```diff
@@ -0,0 +1,19 @@
+    @tags ||= Trends::Tags.new
+  end
+
+  def self.calculate
+    [links, tags].each(&:calculate)
+  end
```
This may be an unfortunate name for a method, especially for something under `app/models/` as `ApplicationRecord` objects do have a `calculate` method, that does something completely different. Maybe `refresh!`? At least ending with an exclamation mark would make sense, I think.

ClearlyClaire reviewed (changes requested):
First off, I still can't understand the need for such a feature. To me, this seems to provide exceedingly little value.
I have to say I am a bit afraid this is part of a trend (ha) to value “engagement” of any kind rather than the quality of the experience itself.
But even as such, even as the “discovery feature” it is pitched in this PR's description, I don't think it achieves much, as it only leads you to resources outside of Mastodon, not to users to follow and interact with.

I am glad that following my complaint this evening, you took steps to limit the most glaring issues with that feature by:
- allowing approving and disapproving individual links without having to approve/disapprove providers (without which admins would have, for instance, to chose between rejecting the BBC as a whole, or accepting their recent infamous transphobic article)
- not counting links behind a Content Warning towards trending links (without which it would be very easy for links usually shared under a Content Warning to trend without it)

Provided a responsible and responsive moderation team, these features should greatly reduce the harm this feature could cause, but fundamentally:
- links are stripped of context, which means both approving and disapproving/criticizing/denouncing/debunking links to said articles would generally count towards trending, therefore, protesting an article while linking to it, debunking disinformation while linking to it, may result in that article trending, without that important context, and may require an active moderation team to remove it
- this is a new target for spammers, probably even more so than hashtags, since it directly links to outside websites; this is somewhat addressed by the need for a priori moderation for any new domain name target, but this doesn't prevent ad pieces on legitimate news outlets from being spammed, and there is the risk of the moderation team drowning behind spam and struggling to approve non-spam providers and links
- while the risk is lessened by CWed links not counting towards trends, not all parts of the fediverse have the same standards for putting content behind a warning, so this may expose people who otherwise curate their timelines with appropriately-CWed content to more un-CWed triggers than otherwise
- conversely, articles on topics that may be worthy of a trigger warning but are nonetheless important and would be trending if they weren't under Content Warnings, won't be trending. I think it is preferable to users being directly exposed to them without a warning, but this nonetheless shoves certain topics outside of the “trending” view

Other minor comments:
- PR description and screenshots are slightly outdated
- I'm not sure all the trending-related database fields make sense to be associated with any and all preview cards; though being nullable and all, the impact is probably minimal

Gargron commented on app/controllers/admin/dashboard_controller.rb:
```diff
@@ -4,7 +4,7 @@ module Admin
     def index
       @system_checks         = Admin::SystemCheck.perform
-      @time_period           = (1.month.ago.to_date...Time.now.utc.to_date)
+      @time_period           = (29.days.ago.to_date...Time.now.utc.to_date)
```
`1.month` is usually 30 days (but not always) but `Time.now.utc` is an extra day, so the total ends up being 31 days. I intended the time period to be a rolling "Last 30 days".

ClearlyClaire replied:
ok

Gargron commented on app/lib/activitypub/activity/create.rb:
```diff
@@ -164,9 +164,14 @@ def postprocess_audience_and_deliver
     @tags.each do |tag|
       status.tags << tag
-      tag.use!(@account, status: status, at_time: status.created_at) if status.public_visibility?
+      tag.update(last_status_at: status.created_at) if tag.last_status_at.nil? || (tag.last_status_at < status.created_at && tag.last_status_at < 12.hours.ago)
```
My thinking was `last_status_at` isn't actually related to trends. Though to be honest I'm not sure what it is related to now, I don't think this value is used by anything. It used to be used by "Last active" sorting in the admin UI but that sorting is not useful at all.

Gargron commented on app/models/form/preview_card_batch.rb:
```diff
@@ -0,0 +1,59 @@
+    preview_cards.update_all(trendable: true)
+  end
+
+  def approve_all!
+    preview_card_providers.each do |provider|
+      authorize(provider, :update?)
+      provider.update(trendable: true, reviewed_at: action_time)
+    end
+  end
+
+  def reject!
+    preview_cards.each { |preview_card| authorize(preview_card, :update?) }
+    preview_cards.update_all(trendable: false)
+  end
+
+  def reject_all!
+    preview_card_providers.each do |provider|
+      authorize(provider, :update?)
+      provider.update(trendable: false, reviewed_at: action_time)
+    end
+  end
```
I guess you mean any selected preview card should have its `trendable` reset to `nil`?

ClearlyClaire replied:
Either `nil` or `true` for `approve_all!` and `false` for `disapprove_all!`, both kind of make sense, but maybe the second one moreso, since you explicitly enabled those ones.

Gargron commented on app/lib/admin/metrics/dimension.rb:
```diff
@@ -7,9 +7,11 @@ class Admin::Metrics::Dimension
-  def self.retrieve(dimension_keys, start_at, end_at, limit)
-    Array(dimension_keys).map { |key| DIMENSIONS[key.to_sym]&.new(start_at, end_at, limit) }.compact
+  def self.retrieve(dimension_keys, start_at, end_at, limit, params)
+    Array(dimension_keys).map { |key| DIMENSIONS[key.to_sym]&.new(start_at, end_at, limit, params[key.to_sym]) }.compact
```
Hm, with `require` it would error out for all dimensions where no special params are necessary.

Gargron commented on app/models/tag_filter.rb:
```diff
@@ -18,35 +13,54 @@ def initialize(params)
+    if ids.empty?
+      Tag.none
+    else
+      Tag.joins("join unnest(array[#{ids.map(&:to_i).join(',')}]::integer[]) with ordinality as x (id, ordering) on tags.id = x.id").order('x.ordering')
```
It is a weird one for sure but I wanted to keep the scope composability of the filter class. We use `page` on the returned relation in the controllers.

Gargron commented on app/models/tag_filter.rb:
```diff
@@ -18,35 +13,54 @@ def initialize(params)
 
   def results
-    scope = Tag.unscoped
+    scope = begin
+      if params[:status] == 'pending_review'
+        Tag.unscoped
+      else
+        trending_scope
+      end
+    end
```
"All" does show tags that are pending, but it does not show *all* pending tags, only the ones that are trending. On the other hand, "Pending", goes over *all* pending tags.

Gargron commented on app/controllers/api/v1/admin/dimensions_controller.rb:
```diff
@@ -17,7 +17,8 @@ def set_dimensions
       params[:end_at],
-      params[:limit]
+      params[:limit],
+      params
```
Right. The first values are always necessary. But I wanted individual classes to decide on their own which params they need, which is why this hand-off is happening here. I'm thinking of it as sort of `const { keys, start_at, end_at, limit, ...other } = params` in JS, not an uncommon pattern.

ThatOneCalculator commented:
cringe :nauseated_face: 
![Screenshot_2021-11-17-03-10-02_2560x1440](https://user-images.githubusercontent.com/44733677/142189774-86185568-dd31-46aa-b40c-3e4c1a27e6c1.png)

shleeable commented:
I like the concept of this kind of feature as a passive insight into the community. Something like hackernews is opt-in, but trending links represents in theory a realtime view of what the community is posting.... and it could be meme'y, or weird, or educational. As long as it's dynamic, you'll get a little view into the hivemind.

This is the same reason I LOVE and adore the federated tmeline. It's a stream of random conscientiousness outside of my bubble.

ClearlyClaire commented on app/workers/scheduler/trending_tags_scheduler.rb:
```diff
@@ -6,6 +6,6 @@ class Scheduler::TrendingTagsScheduler
 
   def perform
-    TrendingTags.update! if Setting.trends
+    Trends.refresh!
```
Why run the worker if the feature is disabled?

Gargron replied:
Because it’s useful information for the admin without being publicly visible and the entire hashtags view is now centered on that sorting

ClearlyClaire replied:
ok I guess hashtags might be interesting, but I can't see why for links, as there is no way for a regular admin to see who posted links or the context

Gargron replied:
Maybe there will be

ClearlyClaire reviewed (commented):
The implementation looks mostly good but I still don't think that feature is a good idea.

Also, if we do merge this feature, it may make sense to have separate settings for trending links and trending tags, I can image some admins being ok with trending hashtags but not trending links, and without separate settings they'd be spammed with review requests anyway.

Gargron committed:
Add link type to preview cards and only trend articles

Change trends review notifications from being sent every 5 minutes to being sent every 2 hours

Change threshold from 5 unique accounts to 15 unique accounts

Gargron committed:
Merge remote-tracking branch 'origin/main' into feature-discover-trending-links

Gargron committed:
Fix tests

ClearlyClaire commented:
I really don't think we should do this.
